### ISSUE TITLE
Tablet assignment on startup is sub-optimal

### ISSUE THREAD

**Issue by andrewglowacki** [apache/accumulo#1049]
When Accumulo initially starts, it appears if all the tablet servers don't come up quickly enough, only a subset of the servers will initially be assigned all the tablets.  This is fine in most cases when that number is close to the total, but when the percentage of online is small, it can cause initial assignment to take 4-5 times longer than normal.

**Comment by ctubbsii** [comment]
Presumably, the master is assigning as fast as it can, to the tservers it knows about so far. We could give the master a configurable amount of seconds to wait for tservers to start up and check in. That could let the master know about more of them before assignments begin.

**Comment by keith-turner** [comment]
I am trying to think of a good function for calculating an initial wait time.  Below is one possibility.  

```
   AT is the actual number of tablet servers
   ET is expected number of tablet servers (configurable by user?  what if AT>ET?)
   EXP is a configurable exponent (configurable by user)
   W is the wait time in seconds

  W = (ET - AT)^EXP / 1000
   
```

The image below is a plot of the wait times (in seconds) for 1,026 expected tablet servers with an exponent of 2.1.

![accumulo-1049](https://user-images.githubusercontent.com/1268739/54835300-3639ac00-4c98-11e9-9fe4-d25e7b5a3284.png)

**Comment by keith-turner** [comment]
I don't think my proposed function works well for small numbers of tablet servers.  Thinking the function should use a ratio of ET vs AT instead of the difference.  For example if I expect 10 and only 1 is online, probably want to wait longer.   The function I proposed would not do this.

**Comment by andrewglowacki** [comment]
Could the master simply wait for a configured % of servers to report similar to how HDFS waits for a configured % of blocks?

**Comment by keith-turner** [comment]
> Could the master simply wait for a configured % of servers to report similar to how HDFS waits for a configured % of blocks?

Yeah something along those lines would work.    Below is another version of the formula.  When the exponent is 1, its a linear ratio.

```
   AT is the actual number of tablet servers
   ET is expected number of tablet servers, configurable by user?
   OFFLINE = max(ET-AT,0)
   EXP is a configurable exponent 
   MAX is a user configurable max wait time in seconds
   W is the wait time in seconds

  W = (OFFLINE/ET)^EXP * MAX
```
This image below is for a max wait time of 1200 secs with exponents 1,1.5,2,and 3.  The Y axis is the computed wait time.  The x axis is the percent of tablet servers offline.

![accumulo-1049-2](https://user-images.githubusercontent.com/1268739/54841537-dba74c80-4ca5-11e9-8c9f-47d69a5495a1.png)

**Comment by ctubbsii** [comment]
We probably want to go ahead and assign the metadata tablets as soon as possible, and then only wait on the rest after a threshold is reached.

**Comment by keith-turner** [comment]
> We probably want to go ahead and assign the metadata tablets as soon as possible, and then only wait on the rest after a threshold is reached.

Could compute a wait time for root, meta, and user tablets separately.  Default config for root could be to not wait.  Default config for meta could be aggressive.

**Comment by milleruntime** [comment]
I vote for the Occam's razor solution, the simplest one that works.

**Comment by keith-turner** [comment]
> I vote for the Occam's razor solution, the simplest one that works.

Maybe thats a property for min servers to do initial assignments and balancing.   While the number of tserver is less than this it does nothing.  Which is a simplification of what @andrewglowacki suggested earlier and I misread, I was thinking of wait time ratios and not server ratios.    Instead of a user configuring two things (an expected server and a ratio) just configure one, the min servers which is the product of the two.

**Comment by EdColeman** [comment]
How would a property account for situations when something other that the default table balancer is used?

In practice, I'll use pssh to start the tservers first, and once they are up and stable I'll start the master.  This gives me the opportunity to check a few tserver logs to validate that things look "normal" before proceeding.

Also, if there is any server affinity for tablets, those servers are most likely ready and waiting for assignments rather that needing to be "rebalanced" after all tservers check in.

**Comment by milleruntime** [comment]
I like the minimum number of tservers because of the simplicity.  Users might prefer a percentage though, since that would mean less adjusting of the property as they bring servers on and offline, better for maintainability.

**Comment by EdColeman** [comment]
The ability to require a number of tservers and a max wait time have been merged into the baseline.

1.9: https://github.com/apache/accumulo/pull/1158
2.0: https://github.com/apache/accumulo/pull/1204
